### PR TITLE
Refactor my-subscribe-button to support full account subscription

### LINKED ISSUE #2389: Being able to subscribe to a user in addition to channel
Hello!

Following a discussion on Mastodon, here is a suggestion about subscription on Peertube: it would be very useful for a user to be able to subscribe to another user, instead of subscribing only to channels one by one.

Indeed, currently the "subscribe" button are displayed next to channel on a user profile, so when you have a Peertube account, it's only possible to subscribe to channels.

If the user create new channel after that (or if the user has numerous channel), it's possible to miss video, and/or we need to check subscription regularly to subscribe to new channel.

But, if one has a mastodon account, it's possible to subscribe to the Peertube account directly using the "activity pub nomination" eg: `lecygnenoir@videos.pair2jeux.tube`

In this case, the mastodon account will receive notification for any video published, regardless the channel used, which is very convenient.

In addition to that, it increase the "subscriber" count for the account instead of increasing counter for each channel  
(By the way it is the only way I know to change this account counter, please do not hesitate to tell me if there is any other way)

I think it would be useful to have this possibility inside Peertube too!

As discussed a possibility would be to add a button to subscribe all existing channels in one click, but I think we could have some problems with this solution:

 - If new channels are created, we need to resubscribe to them
 - If we already subscribe some channel, this button should not resubscribe (or manage duplicated subscription?)
 - Currently when subscribed to a channel, button change to grey and display "unsubscribe", but this new button will need to work differently (proposing to unsubscribe in sub menu? Grey until new channel are created?)

However this is totally opened to discussion, as I just noticed this feature missing but am unaware about the work needed to realize it 😅 

Anyway, thanks again for your wonderful work on Peertube!

### PR DESCRIPTION
**Pull request by rigelk**
resolves #2389

Screenshot_2020-01-08 Account video channels - PeerTube

still lacking:
- [x] proper sync with other `my-subscribe-button`s on the same view
- [x] display of how many channels are not subscribed yet

side changes:
- uppercase account/channel nav
- single/plural for "view(s)"
- single/plural for "subscriber(s)"

### CODE REVIEW

**Comment by DontUseGithub** [comment]
Regarding the button, UI-wise I'd recommend keeping only the word "Subscribe" + the dropdown icon. And once clicked, the user will see all the options, one of which will be "Subscribe to all channels..."

**Comment by rigelk** [comment]
> Regarding the button, UI-wise I'd recommend keeping only the word "Subscribe" + the dropdown icon. And once clicked, the user will see all the options, one of which will be "Subscribe to all channels..."

I don't agree. The user won't understand the action beforehand, nor have information about how many channels are left to subscribe to.

**Comment by DontUseGithub** [comment]
To me it looks consistent. If you visit an _acount page_ like this one:

peertube2.cpy.re/**accounts**/root/video-channels 

and see a subscribe button, then you know you're subscribing to the account and everything that account does.

And if you visit a _channel page_ like this one:

peertube2.cpy.re/**video-channels**/a75cbdf4-acf2-45c0-a491-e9b28939f8db/videos

and see a subscribe button, then you know you're subscribing to that channel.

Maybe to make these things more obvious, we could emphasize somehow which page is an account page and which one is a channel page, not everyone is looking at the urls and nothing besides the url at the moment differentiates between these two types of pages (text-wise).

Maybe there's a better approach to this, but in my opinion a button with 7 words inside it doesn't look very good (UI-wise). Hopefully this doesn't sound rude, if it doesn't I'm really sorry.

**Comment by rigelk** [comment]
> To me it looks consistent. If you visit an _acount page_ like this one:
> 
> peertube2.cpy.re/**accounts**/root/video-channels
> 
> and see a subscribe button, then you know you're subscribing to the account and everything that account does.

Except you won't be. PeerTube doesn't support following accounts - it just supports following channels. That means following an account means actually following each of its channels at the time of the subscription. If a new channel is created, you won't be notified of content put in it.

Yes, it is a simplification. We could have made following accounts possible in the backend, but it is arguably a more complex change to the codebase than what this PR does.

> Maybe to make these things more obvious, we could emphasize somehow which page is an account page and which one is a channel page, not everyone is looking at the urls and nothing besides the url at the moment differentiates between these two types of pages (text-wise).

The "Created by account@host" differentiates channels, for instance. The navigation bar below also changes.

> Hopefully this doesn't sound rude, if it does I'm really sorry, that's not my intentions.

No offense taken! Thank you for caring :slightly_smiling_face:

**Comment by rigelk** [comment]
@Chocobozzz I increased the size of the button

Screenshot_2020-01-09 Account video channels - PeerTube

WDYT?

File: client/src/app/shared/user-subscription/user-subscription.service.ts [hunk 1]
```diff
@@ -69,10 +97,46 @@ export class UserSubscriptionService {
...
+                      filter(existsResult => existsResult[ nameWithHost ] !== undefined),
+                      map(existsResult => existsResult[ nameWithHost ])
+                    )
+
+    this.myAccountSubscriptionCacheObservable[ nameWithHost ] = obs
+    return obs
+  }
+
   doesSubscriptionExist (nameWithHost: string) {
+    logger('Running subscription check')
```
Comment (Chocobozzz) [inline client/src/app/shared/user-subscription/user-subscription.service.ts]: `for ${nameWithHost}`

File: client/src/app/shared/user-subscription/subscribe-button.component.ts [hunk 2]
```diff
@@ -64,15 +78,30 @@ export class SubscribeButtonComponent implements OnInit {
...
         () => {
-          this.subscribed = true
-
           this.notifier.success(
-            this.i18n('Subscribed to {{nameWithHost}}. You will be notified of all their new videos.',
-              { nameWithHost: this.videoChannel.displayName }
-            ),
+            this.account
+              ? this.i18n(
+                  'Subscribed to all current channels of {{nameWithHost}}. ' +
```
[the remark this answers is not in the document]
Comment (rigelk) [reply client/src/app/shared/user-subscription/subscribe-button.component.ts]: aaah, that explains your recent change to the message. I had to rebase but didn't understand your change.

File: client/src/app/shared/user-subscription/subscribe-button.component.ts [hunk 3]
```diff
@@ -64,15 +78,30 @@ export class SubscribeButtonComponent implements OnInit {
   }
 
   localSubscribe () {
-    this.userSubscriptionService.addSubscription(this.channelHandle)
+    const observableBatch: any = []
+
+    this.videoChannels
+      .filter(videoChannel => this.subscribeStatus(false).includes(this.getChannelHandler(videoChannel)))
+      .forEach(videoChannel => observableBatch.push(
```
[the remark this answers is not in the document]
Comment (Chocobozzz) [reply client/src/app/shared/user-subscription/subscribe-button.component.ts]: does not seem to be resolved?

File: client/src/app/shared/user-subscription/subscribe-button.component.ts [hunk 4]
```diff
@@ -107,7 +140,49 @@ export class SubscribeButtonComponent implements OnInit {
...
+    for (const [key, value] of this.subscribed.entries()) {
+      if (value === subscribed) accumulator.push(key)
+    }
+    return accumulator
+  }
+
+  private loadSubscribedStatus () {
+    if (!this.isUserLoggedIn()) return
+
+    this.videoChannels.map(videoChannel => {
```
Comment (Chocobozzz) [inline client/src/app/shared/user-subscription/subscribe-button.component.ts]: just use a for of

File: client/src/app/shared/user-subscription/subscribe-button.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rigelk) [reply client/src/app/shared/user-subscription/subscribe-button.component.ts]: I… thought I had ><

[the remark this answers is not in the document]
Comment (rigelk) [reply client/src/app/shared/user-subscription/subscribe-button.component.ts]: Ah, I did it for the addSubscription and forgot the deleteSubscription.

**Pull request merged**